File: 1008.construct-binary-search-tree-from-preorder-traversal.cpp
/*
 * @lc app=leetcode id=1008 lang=cpp
 *
 * [1008] Construct Binary Search Tree from Preorder Traversal
 */
#include <vector>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    TreeNode *bstFromPreorder(vector<int> &preorder)
    {
        TreeNode *root = new TreeNode(preorder[0]);
        auto cur = root;
        vector<TreeNode *> node_stack = {root};
        for (int i = 1, width = preorder.size(); i != width; ++i)
        {
            int cur_val = preorder[i];
            if (node_stack.back()->val > cur_val)
            {
                cur->left = new TreeNode(cur_val);
                cur = cur->left;
                node_stack.push_back(cur);
            }
            else
            {
                while (!node_stack.empty() && node_stack.back()->val < cur_val)
                {
                    cur = node_stack.back();
                    node_stack.pop_back();
                }
                cur->right = new TreeNode(cur_val);
                cur = cur->right;
                node_stack.push_back(cur);
            }
        }
        return root;
    }
};
// @lc code=end

File: 1008.construct-binary-search-tree-from-preorder-traversal.py
#
# @lc app=leetcode id=1008 lang=python3
#
# [1008] Construct Binary Search Tree from Preorder Traversal
#
from typing import *


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        root = TreeNode(preorder[0])
        cur = root
        node_stack = [root]
        for i in range(1, len(preorder)):
            cur_val = preorder[i]
            if node_stack[-1].val > cur_val:
                cur.left = TreeNode(cur_val)
                cur = cur.left
                node_stack.append(cur)
            else:
                while len(node_stack) and node_stack[-1].val < cur_val:
                    cur = node_stack[-1]
                    node_stack.pop()
                cur.right = TreeNode(cur_val)
                cur = cur.right
                node_stack.append(cur)
        return root

# @lc code=end
